### PR TITLE
supplementary number #212 for authorized login services

### PR DESCRIPTION
**Pull request by DaVinci9196**
completion service <com.google.android.gms.auth.api.identity.service.signin.START>

### CODE REVIEW

File: play-services-basement/src/main/java/com/google/android/gms/common/internal/safeparcel/AbstractSafeParcelable.java [hunk 1]
```diff
@@ -21,4 +23,13 @@ public static <T extends AbstractSafeParcelable> SafeParcelableCreatorAndWriter<
     public int describeContents() {
         return 0;
     }
+
+    public byte[] toByteArray() {
```
Comment (mar-v-in) [inline play-services-basement/src/main/java/com/google/android/gms/common/internal/safeparcel/AbstractSafeParcelable.java:27]: Please use `SafeParcelableSerializer.serializeToBytes()` instead.

File: play-services-auth/src/main/java/com/google/android/gms/auth/api/identity/BeginSignInRequest.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/*
+ * SPDX-FileCopyrightText: 2023 microG Project Team
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.google.android.gms.auth.api.identity;
+
+import org.microg.safeparcel.AutoSafeParcelable;
+
+public class BeginSignInRequest extends AutoSafeParcelable {
```
Comment (mar-v-in) [inline play-services-auth/src/main/java/com/google/android/gms/auth/api/identity/BeginSignInRequest.java:10]: Please make use of the new safe-parcel-processor wherever easily possible.
1. Ensure that `annotationProcessor project(':safe-parcel-processor')` is added to `dependencies` of corresponding `build.gradle`.
2. Make the class extend `AbstractSafeParcelable` instead of `AutoSafeParcelable`
3. Add `@SafeParcelable.Class` annotiation to the class
4. Make sure the `CREATOR` is defined as `public static final SafeParcelableCreatorAndWriter<XXX> CREATOR = findCreator(XXX.class);`
5. Add `writeToParcel` to the class:
  ```java
  @Override
  public void writeToParcel(@NonNull Parcel dest, int flags) {
      CREATOR.writeToParcel(this, dest, flags);
  }
  ```
6. If necessary, make private fields package-visible or provide a getter (referenced via `getterName`) and constructor parameter for them, tagged with `@Param`.
7. If necessary, tag the constructor to be used when deserializing with `@Constructor` (if no constructor is declared, the default constructor will be used).

Using this system severely reduces the use of reflection at runtime, which can significantly improve performance as it allows for better optimizations.

File: play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt [hunk 3]
```diff
@@ -197,18 +200,37 @@ class AuthSignInActivity : AppCompatActivity() {
...
+            SignInAccount().apply {
                 email = googleSignInAccount.email ?: account?.name
                 this.googleSignInAccount = googleSignInAccount
-                userId = googleSignInAccount.id ?: getSystemService<AccountManager>()?.getUserData(account, "GoogleUserId")
-            })
+                userId = googleSignInAccount.id ?: getSystemService<AccountManager>()?.getUserData(
+                    account,
+                    AuthConstants.GOOGLE_USER_ID
+                )
+            }.let {
```
Comment (mar-v-in) [inline play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt:215]: Why did you change this to a `let` expression?

File: play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt [hunk 4]
```diff
@@ -197,18 +200,37 @@ class AuthSignInActivity : AppCompatActivity() {
...
+                data.putExtra(AuthConstants.SIGN_IN_ACCOUNT, it)
+            }
+            SignInCredential().apply {
+                email = googleSignInAccount.email
+                displayName = googleSignInAccount.displayName
+                familyName = googleSignInAccount.familyName
+                givenName = googleSignInAccount.givenName
+                idToken = googleSignInAccount.idToken
+            }.let {
```
Comment (mar-v-in) [inline play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt:224]: Same question here.

File: play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt [hunk 5]
```diff
@@ -197,18 +200,37 @@ class AuthSignInActivity : AppCompatActivity() {
...
+                bundle.putByteArray(AuthConstants.SIGN_IN_CREDENTIAL, it.toByteArray())
+                bundle.putByteArray(AuthConstants.STATUS, Status.SUCCESS.toByteArray())
```
Comment (mar-v-in) [inline play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt:226]: This line does not even depend on the `SignInCredential` object you just created, so it makes little sense to have it in the same `let` block.

File: play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import org.microg.gms.BaseService
+import org.microg.gms.common.GmsService
+
+const val TAG = "AssistedSignInService"
+
+class AssistedSignInService : BaseService(TAG, GmsService.IDENTITY_SIGN_IN) {
+
+    override fun handleServiceRequest(callback: IGmsCallbacks, request: GetServiceRequest, service: GmsService) {
+        val connectionInfo = ConnectionInfo()
+        connectionInfo.features = arrayOf(
```
Comment (mar-v-in) [inline play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt:22]: This seems to be the feature set already defined in `org.microg.gms.auth.credentials.FEATURES`, so better don't declare it another time, so we don't have to update at two locations.

File: play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
+/*
+ * SPDX-FileCopyrightText: 2023 microG Project Team
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.microg.gms.auth.credentials.assistedsignin
```
Comment (mar-v-in) [inline play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt:6]: Please use package `org.microg.gms.auth.credentials.identity` and name `IdentitySignInService` to match AIDL, public API and existing naming schemes.

**Review by mar-v-in — CHANGES_REQUESTED** [review]

File: play-services-basement/src/main/java/com/google/android/gms/common/internal/safeparcel/AbstractSafeParcelable.java (hunk 1 above)
Comment (DaVinci9196) [reply play-services-basement/src/main/java/com/google/android/gms/common/internal/safeparcel/AbstractSafeParcelable.java]: applied

File: play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt (hunk 3 above)
Comment (DaVinci9196) [reply play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt]: it changed

File: play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt (hunk 5 above)
Comment (DaVinci9196) [reply play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt]: it changed

File: play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt (hunk 6 above)
Comment (DaVinci9196) [reply play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt]: you are right, I found it is the same

File: play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt (hunk 7 above)
Comment (DaVinci9196) [reply play-services-core/src/main/kotlin/org/microg/gms/auth/credentials/assistedsignin/AssistedSignInService.kt]: applied

File: play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt (hunk 4 above)
Comment (DaVinci9196) [reply play-services-core/src/main/kotlin/org/microg/gms/auth/signin/AuthSignInActivity.kt]: it changed

File: play-services-auth/src/main/java/com/google/android/gms/auth/api/identity/BeginSignInRequest.java (hunk 2 above)
Comment (DaVinci9196) [reply play-services-auth/src/main/java/com/google/android/gms/auth/api/identity/BeginSignInRequest.java:10]: I don’t understand it very well here. AutoSafeParcelable inherits from AbstractSafeParcelable and uses SafeParcelableCreatorAndWriter.

Comment (mar-v-in) [reply play-services-auth/src/main/java/com/google/android/gms/auth/api/identity/BeginSignInRequest.java:10]: `AutoSafeParcelable` is using reflection. Reflection isn't good for performance as it can't be optimized. When using the system based on the annotation processor, as explained above, reflection is not used during (de)serialization. I outlined how to use annotation processor above, you can also take a look at other classes that directly inherit from `AbstractSafeParcelable` without using `AutoSafeParcelable`.

Comment (DaVinci9196) [reply play-services-auth/src/main/java/com/google/android/gms/auth/api/identity/BeginSignInRequest.java]: Oh oh, I changed it
